source/workspaced/com/ccdb.d

```diff
@@ -0,0 +1,287 @@
+
+		string jsonString = cast(string) assumeUnique(read(filename));
+		auto json = parseJSON(jsonString);
+		// clang db can be quite large (e.g. 100 k lines of JSON data on large projects)
+		// we release memory when possible to avoid having at the same time more than
+		// two represention of the same data
+		jsonString = null;
+
+		HashSet!string imports;
+		HashSet!string stringImports;
+		HashSet!string fileImports;
+		HashSet!string versions;
+		HashSet!string debugVersions;
+
+		json.array
+			.map!(jv => CompileCommand.fromJson(jv))
+			.filter!(cc => cc.isValid)
+			.each!(cc =>
+					cc.feedOptions(imports, stringImports, fileImports, versions, debugVersions)
+			);
+
+		_importPaths = imports[].array;
+		_stringImportPaths = stringImports[].array;
+		_importFiles = fileImports[].array;
+		_versions = versions[].array;
+		_debugVersions = debugVersions[].array;
+	}
+
+	/// Lists all import paths
+	string[] imports() @property nothrow
+	{
+		return _importPaths;
+	}
+
+	/// Lists all string import paths
+	string[] stringImports() @property nothrow
+	{
+		return _stringImportPaths;
+	}
+
+	/// Lists all import paths to files
+	string[] fileImports() @property nothrow
+	{
+		return _importFiles;
+	}
+
+	/// Lists the currently defined versions
+	string[] versions() @property nothrow
+	{
+		return _versions;
+	}
+
+	/// Lists the currently defined debug versions (debug specifications)
+	string[] debugVersions() @property nothrow
+	{
+		return _debugVersions;
+	}
+
+private:
+
+	string[] _importPaths, _stringImportPaths, _importFiles, _versions, _debugVersions;
+}
+
+private struct CompileCommand
+{
+	string directory;
+	string file;
+	string[] args;
+	string output;
+
+	static CompileCommand fromJson(JSONValue json)
+	{
+		import std.algorithm : map;
+		import std.array : array;
+
+		CompileCommand cc;
+
+		cc.directory = json["directory"].str;
+		cc.file = json["file"].str;
+
+		if (auto args = "arguments" in json)
+		{
+			cc.args = args.array.map!(jv => jv.str).array;
+		}
+		else if (auto cmd = "command" in json)
+		{
+			cc.args = unescapeCommand(cmd.str);
+		}
+		else
+		{
+			throw new Exception(
+				"Either 'arguments' or 'command' missing from Clang compilation database");
+		}
+
+		if (auto o = "output" in json)
+		{
+			cc.output = o.str;
+		}
+
+		return cc;
+	}
+
+	@property bool isValid() const
+	{
+		import std.algorithm : endsWith;
+
+		if (args.length <= 1)
+			return false;
+		if (!file.endsWith(".d"))
+			return false;
+		return true;
+	}
+
+	void feedOptions(
+		ref HashSet!string imports,
+		ref HashSet!string stringImports,
+		ref HashSet!string fileImports,
+		ref HashSet!string versions,
+		ref HashSet!string debugVersions)
+	{
+		import std.algorithm : startsWith;
+
+		enum importMark = "-I=";
+		enum stringImportMark = "-J=";
+		enum fileImportMark = "-i=";
+		enum versionMark = "-version=";
+		enum debugMark = "-debug=";
+
+		foreach (arg; args)
+		{
+			if (arg.startsWith(importMark))
```

Comment:
you could use the `startsWith` overload taking in multiple needles and switch on that, e.g.
```d
auto mark = arg.startsWith(importMark, stringImportMark, fileImportMark, versionMark, debugMark);
// mark is 0 if none, 1=importMark, 2=stringImportMark, 3=fileImportMark, 4=versionMark, 5=debugMark
```

Reply:
I didn't know this overload!